### PR TITLE
Enable subscription via Stripe on independent apk

### PR DESCRIPTION
**Pull request by ua741**

### CODE REVIEW

File: lib/models/subscription.dart [hunk 1]
```diff
@@ -52,7 +60,7 @@ class Subscription {
   }
 
   Map<String, dynamic> toMap() {
-    return {
+    var map = <String, dynamic>{
```
Comment (vishnukvmd) [inline lib/models/subscription.dart:63]: We use `final` by default.

File: lib/models/subscription.dart [hunk 2]
```diff
@@ -62,11 +70,14 @@ class Subscription {
       'price': price,
       'period': period,
```
Comment (vishnukvmd) [inline lib/models/subscription.dart:71]: The following line can be included here I think:
```
'attributes': attributes?.toJson()
```

File: lib/ui/payment/payment_web_page.dart [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+  final String actionType;
+
+  const PaymentWebPage({Key key, this.planId, this.actionType})
+      : super(key: key);
+
+  @override
+  State<StatefulWidget> createState() => _PaymentWebPage();
+}
+
+class _PaymentWebPage extends State<PaymentWebPage> {
```
Comment (vishnukvmd) [inline lib/ui/payment/payment_web_page.dart:32]: Convention is to suffix the state with `State`. So `_PaymentWebPageState`.

File: lib/ui/payment/payment_web_page.dart [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+  final _logger = Logger("PaymentWebPage");
+  UserService userService = UserService.instance;
```
Comment (vishnukvmd) [inline lib/ui/payment/payment_web_page.dart:34]: Whatever can be `final` could be `final`?

File: lib/ui/payment/payment_web_page.dart [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+      });
+    });
+    if (Platform.isAndroid && kDebugMode) {
+      AndroidInAppWebViewController.setWebContentsDebuggingEnabled(true);
+    }
+    _dialog = createProgressDialog(context, "please wait...");
+    super.initState();
+  }
+
+  Uri _getPaymentUrl(String paymentToken) {
```
Comment (vishnukvmd) [inline lib/ui/payment/payment_web_page.dart:56]: We usually try to order methods based on visibility such that public methods are defined first.

File: lib/ui/payment/payment_web_page.dart [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+                      await _dialog.show();
+                    }
+                  },
+                  onLoadError: (controller, navigationAction, code, msg) async {
+                    if (_dialog.isShowing()) {
+                      await _dialog.hide();
+                    }
+                  },
+                  onLoadHttpError: (controller, navigationAction, code, msg) async {
+                    _logger.info("onHttpError with $code and msg = $msg");
```
Comment (vishnukvmd) [inline lib/ui/payment/payment_web_page.dart:126]: Should we log the error to Sentry?

File: lib/ui/payment/payment_web_page.dart [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+    _dialog.hide();
+    super.dispose();
+  }
+
+  // show dialog to handle accidental back press.
+  Future<bool> _buildPageExitWidget(BuildContext context) {
+    return showDialog(
+        context: context,
+        builder: (context) => AlertDialog(
+                title: Text('Are you sure you want to exit?'),
```
Comment (vishnukvmd) [inline lib/ui/payment/payment_web_page.dart:152]: all small case 😬

File: lib/ui/payment/payment_web_page.dart [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+                      child: Text('no'),
+                      onPressed: () => Navigator.of(context).pop(false)),
+                ]));
+  }
+
+  bool _isPaymentActionComplete(Uri loadingUri) {
+    return loadingUri.toString().startsWith(kWebPaymentRedirectUrl);
+  }
+
+  Future<void> handlePaymentResponse(Uri uri) async {
```
Comment (vishnukvmd) [inline lib/ui/payment/payment_web_page.dart:167]: Could be a private method?

File: lib/ui/payment/payment_web_page.dart [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+    var queryParams = uri.queryParameters;
+    var paymentStatus = uri.queryParameters['status'] ?? '';
+    _logger.fine('handle payment response with status $paymentStatus');
+    if (paymentStatus == 'success') {
+      await _handlePaymentSuccess(queryParams);
+    } else if ('fail' == paymentStatus) {
```
Comment (vishnukvmd) [inline lib/ui/payment/payment_web_page.dart:173]: Any reason for flipping the ordering?

File: lib/ui/payment/payment_web_page.dart [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+                content:
+                    Text("unfortunately your payment failed due to $reason"),
+                actions: <Widget>[
+                  TextButton(
+                      child: Text('ok'),
+                      onPressed: () {
+                        Navigator.of(context).pop('dialog');
+                      }),
+                ]));
+    Navigator.of(context).pop(true);
```
Comment (vishnukvmd) [inline lib/ui/payment/payment_web_page.dart:198]: Does this not pop the dialog?

File: lib/ui/payment/payment_web_page.dart [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+    var checkoutSessionID = queryParams['session_id'] ?? '';
+    await _dialog.show();
+    try {
+      var response = await billingService.verifySubscription(
+          widget.planId, checkoutSessionID,
+          paymentProvider: kStripe);
+      await _dialog.hide();
+      if (response != null) {
+        var content = widget.actionType == 'buy'
+            ? 'you subscription purchase was successful'
```
Comment (vishnukvmd) [inline lib/ui/payment/payment_web_page.dart:212]: `your purchase was successful`

File: lib/ui/payment/payment_web_page.dart [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+            : 'your subscription update was successful';
```
Comment (vishnukvmd) [inline lib/ui/payment/payment_web_page.dart:213]: `your subscription was updated successfully`

File: lib/ui/payment/stripe_subscription_page.dart [hunk 13]
```diff
@@ -0,0 +1,420 @@
...
+
+  // _filterPlansForUI is used for initializing initState & plan toggle states
+  Future<void> _filterStripeForUI() async {
+    final billingPlans = await _billingService.getBillingPlans();
+    _freePlan = billingPlans.freePlan;
+    _plans = billingPlans.plans.where((plan) {
+      if (plan.stripeID == null || plan.stripeID.isEmpty) {
+        return false;
+      }
+      var isYearlyPlan = plan.period == 'year';
```
Comment (vishnukvmd) [inline lib/ui/payment/stripe_subscription_page.dart:87]: `final`

File: lib/ui/payment/payment_web_page.dart (hunk 6 above)
Comment (ua741) [reply lib/ui/payment/payment_web_page.dart:126]: Failure could be also due to bad internet.. Don't see the need of logging that to sentry.. 
But can still change it. We can turn if off in the future if it becomes noisy (🤞)

File: lib/ui/payment/payment_web_page.dart (hunk 10 above)
Comment (ua741) [reply lib/ui/payment/payment_web_page.dart:198]: This will take user to back page. Err dialog is popped on 'ok' press.

**Pull request merged**